pennylane/templates/state_preparations/state_prep_mps.py

```diff
@@ -15,24 +15,233 @@
+        (since the dimension of a qubit is :math:`2`), while the other dimensions must be powers of two.
+
+        The following example shows a valid MPS input containing four tensors with
+        dimensions :math:`[(2,2), (2,2,4), (4,2,2), (2,2)]` which satisfy the criteria described above.
+
+        .. code-block::
+
+            mps = [
+                np.array([[0.0, 0.107], [0.994, 0.0]]),
+                np.array(
+                    [
+                        [[0.0, 0.0, 0.0, -0.0], [1.0, 0.0, 0.0, -0.0]],
+                        [[0.0, 1.0, 0.0, -0.0], [0.0, 0.0, 0.0, -0.0]],
+                    ]
+                ),
+                np.array(
+                    [
+                        [[-1.0, 0.0], [0.0, 0.0]],
+                        [[0.0, 0.0], [0.0, 1.0]],
+                        [[0.0, -1.0], [0.0, 0.0]],
+                        [[0.0, 0.0], [1.0, 0.0]],
+                    ]
+                ),
+                np.array([[-1.0, -0.0], [-0.0, -1.0]]),
+            ]
+    """
+
+    _validate_mps_shape(mps)
+
+    mps.copy()
+    mps[0] = mps[0].reshape((1, *mps[0].shape))
+    mps[-1] = mps[-1].reshape((*mps[-1].shape, 1))
+
+    is_right_canonical = True
+    for tensor in mps[1:-1]:
+        # Right-canonical definition
+        input_matrix = qml.math.tensordot(tensor, tensor.conj(), axes=([1, 2], [1, 2]))
+        if not qml.math.allclose(input_matrix, qml.math.eye(tensor.shape[0])):
+            is_right_canonical = False
+            break
+
+    if is_right_canonical:
+        mps[0] = mps[0][0]
+        mps[-1] = mps[-1][:, :, 0]
+        return mps
+
+    d_shapes = []
+    for tensor in mps[1:-1]:
+        d_shapes += tensor.shape
+
+    max_bond_dim = qml.math.max(d_shapes)
+
+    n_sites = len(mps)
+    output_mps = [None] * n_sites
+
+    # Procedure analogous to the left-canonical conversion but starting from the right and storing the Vd,
+    # where Vd is the right matrix in the Singular Value Decomposition (SVD)
+    for i in range(n_sites - 1, 0, -1):
+        chi_left, d, chi_right = mps[i].shape
+        input_matrix = mps[i].reshape(chi_left, d * chi_right)
+
+        u_matrix, s_diag, vd_matrix = qml.math.linalg.svd(input_matrix, full_matrices=False)
+
+        # Truncate SVD components if needed
+        chi_new = min(int(max_bond_dim), len(s_diag))
+        u_matrix = u_matrix[:, :chi_new]
+        s_diag = s_diag[:chi_new]
+        vd_matrix = vd_matrix[:chi_new, :]
+
+        # Store Vd reshaped as an MPS tensor in the output MPS
+        output_mps[i] = vd_matrix.reshape(chi_new, d, chi_right)
+
+        # Contract U with diag(S) and merge it with the preceding MPS tensor, preserving the canonical structure
+        mps[i - 1] = qml.math.tensordot(
+            mps[i - 1], u_matrix @ qml.math.diag(s_diag), axes=([2], [0])
+        )
+
+    output_mps[0] = mps[0][0]
+    output_mps[-1] = output_mps[-1][:, :, 0]
```

Comment:
same as above, does this work with JAX arrays?

Reply:
Added test